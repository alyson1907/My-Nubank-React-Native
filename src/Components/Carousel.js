import React, { useState, useEffect } from 'react'
import { Animated, View, Text, ART, Dimensions } from 'react-native'
import { ScrollView } from 'react-native-gesture-handler'
import styles from './Styles/Carousel'

const Carousel = (props) => {
  const hasChildren = !!props.children
  const numberOfItems = hasChildren && props.children.length || 0

  const [selectedBullet, setSelectedBullet] = useState(0)
  const [width, setWidth] = useState(Dimensions.get('window').width)

  const handleBullet = e => {
    const getItemIndex = e => {
      const contentWidth = width
      // const { width: contentWidth } = e.nativeEvent.contentSize
      const { x: contentOffset } = e.nativeEvent.contentOffset

      const pageWidth = parseInt(contentWidth / numberOfItems)
      const currentPage = Math.round(contentOffset / pageWidth)

      return currentPage
    }
    const currentBullet = getItemIndex(e)
    // TODO add Animation to bullet change
    setSelectedBullet(currentBullet)
  }

  const getPageBullets = (numberOfItems, selectedBullet) => {
    return (
      <View style={styles.bulletsContainer}>
        {hasChildren && props.children.map((child, idx) => {
          return idx === selectedBullet ? (<View key={idx} style={styles.selectedBullet} />) : (<View key={idx} style={styles.bullet} />)
        })}
      </View>
    )
  }

  return (
    <View style={props.style}>
      <ScrollView horizontal={true}
        showsHorizontalScrollIndicator={false}
        contentContainerStyle={{ width: `${100 * numberOfItems}%`, justifyContent: 'space-around' }}
        onContentSizeChange={(width, height) => setWidth(width)}
        decelerationRate="normal"
        pagingEnabled
        onScroll={handleBullet}
      >
        {props.children}
      </ScrollView>
      {props.bullets && getPageBullets(numberOfItems, selectedBullet)}
    </View>
  )
}

export default Carousel
