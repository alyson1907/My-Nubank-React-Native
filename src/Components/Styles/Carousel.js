import { StyleSheet } from 'react-native'
import Colors from '../../Colors'

const styles = StyleSheet.create({
  bulletsContainer: {
    flexDirection: 'row',
    justifyContent: 'center'
  },

  bullet: {
    width: 10,
    height: 10,
    backgroundColor: Colors.primaryAlternative,
    borderRadius: 10,
    marginHorizontal: 8,
    marginTop: 8
  },

  selectedBullet: {
    width: 11,
    height: 11,
    backgroundColor: Colors.secondary,
    borderRadius: 10,
    marginHorizontal: 8,
    marginTop: 8
  }
})

export default styles
