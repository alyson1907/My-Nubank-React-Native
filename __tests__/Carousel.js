import 'react-native'
import React from 'react'
import { View } from 'react-native'
import Carousel, { getPageBullets } from '../src/Components/Carousel.js'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  const tree = renderer.create(<Carousel />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders carousel bullets with 3 items correctly', () => {
  // 3 items, the 1 is selected
  const tree = renderer.create(getPageBullets(3, 1)).toJSON()
  expect(tree).toMatchSnapshot()
})
