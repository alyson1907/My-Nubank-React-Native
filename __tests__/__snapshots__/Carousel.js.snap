// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders carousel bullets with 3 items correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ac14d9",
        "borderRadius": 10,
        "height": 10,
        "marginHorizontal": 8,
        "marginTop": 8,
        "width": 10,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#eee",
        "borderRadius": 10,
        "height": 11,
        "marginHorizontal": 8,
        "marginTop": 8,
        "width": 11,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#ac14d9",
        "borderRadius": 10,
        "height": 10,
        "marginHorizontal": 8,
        "marginTop": 8,
        "width": 10,
      }
    }
  />
</View>
`;

exports[`renders correctly 1`] = `
<View>
  <RCTScrollView
    collapsable={false}
    contentContainerStyle={
      Object {
        "justifyContent": "space-around",
        "width": "0%",
      }
    }
    decelerationRate="normal"
    horizontal={true}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onScroll={[Function]}
    pagingEnabled={true}
    showsHorizontalScrollIndicator={false}
  >
    <View />
  </RCTScrollView>
</View>
`;
